libs/ml/examples/word2vec_utilities.hpp

```diff
@@ -0,0 +1,215 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "math/clustering/knn.hpp"
+#include "math/tensor.hpp"
+#include "ml/dataloaders/word2vec_loaders/sgns_w2v_dataloader.hpp"
+
+#include <cstring>
+#include <dirent.h>
+#include <fstream>
+#include <sstream>
+#include <vector>
+
+namespace fetch {
+namespace ml {
+namespace examples {
+
+template <class TensorType>
+std::vector<std::pair<typename TensorType::SizeType, typename TensorType::Type>> GetWordIDAnalogies(
+    TensorType const &embeddings, typename TensorType::SizeType const &word1,
+    typename TensorType::SizeType const &word2, typename TensorType::SizeType const &word3,
+    typename TensorType::SizeType k)
+{
+  using SizeType = typename TensorType::SizeType;
+  using DataType = typename TensorType::Type;
+
+  // get word vectors for word_ids
+  TensorType word1_vec = embeddings.Slice(word1, 1).Copy();
+  TensorType word2_vec = embeddings.Slice(word2, 1).Copy();
+  TensorType word3_vec = embeddings.Slice(word3, 1).Copy();
+
+  //  word1_vec /= fetch::math::L2Norm(word1_vec);
```

Comment:
you don't normalise these anymore? I thought that was important (to ensure you're really just comparing the angle between entities in embeddings space, rather than absolute distance, which may be less relevant)?

In any case, left in commented out code